### PR TITLE
Implement auto-recovery

### LINKED ISSUE #387: activatable script/top level service on failure of "boot" one
On a normal PC when the boot fails you can at least read something on the display (even if not ideal and usually the failing service is way at the top and scrolled past already at the end when all boot services are stopped) and interact with the usual [`Choose: (r)eboot, r(e)covery, re(s)tart boot sequence, (p)ower off? `](https://github.com/davmac314/dinit/blob/116633b/src/dinit.cc#L802) options; however in my use case (a random Android phone with vendor kernel) that's not a practical option due to typically non-functional VT framebuffer (not to mention the need to use a USB keyboard via OTG, so I wrote this pretty bad patch which does work wonderfully for my needs:
```diff
--- a/src/dinit.cc
+++ b/src/dinit.cc
@@ -193,6 +193,9 @@ namespace {
     // These need to be at namespace scope to prevent causing stack allocations when using them:
     constexpr auto shutdown_exec = literal(SBINDIR) + "/" + SHUTDOWN_PREFIX + "shutdown";
     constexpr auto error_exec_sd = literal("Error executing ") + shutdown_exec + ": ";
+    // HYBRIS HACK: Panic command handler in case dinit fails to boot the system
+    constexpr auto panic_exec = literal(SBINDIR) + "/" + "dinit-panic";
+    constexpr auto error_exec_panic = literal("Error executing ") + panic_exec + ": ";
 }
 
 // Options handled in dinit_main
@@ -700,7 +703,17 @@ int dinit_main(int argc, char **argv)
             // Services all stopped but there was no shutdown issued. Inform user, wait for ack, and
             // re-start boot sequence.
             sync(); // Sync to minimise data loss if user elects to power off / hard reset
-            confirm_restart_boot();
+            // HYBRIS HACK: There's likely no working VT for user input in confirm_restart_boot()
+            log_msg_end(" Will run dinit-panic or reboot.");
+            flush_log();
+            sync();
+
+            execl(panic_exec.c_str(), panic_exec.c_str(), nullptr);
+            log(loglevel_t::ERROR, error_exec_panic, strerror(errno));
+            flush_log();
+            sync();
+            shutdown_type = shutdown_type_t::REBOOT;
+
             if (services->count_active_services() != 0) {
                 // Recovery service started
                 goto run_event_loop;
```

Would anything like that be interesting here for upstream? To be clear this more likely would take the form of a `boot`-adjacent "top level service" like `recovery` or maybe `panic` which can be activated via configuration around https://github.com/davmac314/dinit/blob/116633b/src/dinit.cc#L721 instead of calling the hardcoded `confirm_restart_boot()` function assuming working display/keyboard input.

For further context https://github.com/JamiKettunen/chimera-libhybris/blob/8adfd6d/overlays/base/usr/bin/dinit-panic is how I use the feature to setup a USB network between the device side and a host PC it's attached to so I can `ssh` in to inspect and fix "the damage" or whatever may have gone wrong and prevented a normal boot from finishing :)

### PR DESCRIPTION
**Pull request by JamiKettunen**
In some setups it makes sense to launch dinit with `--auto-recovery` and handle boot failures purely via the `recovery` service always. Likely missing some stuff etc but this implements the feature in a hopefully proper way.

Fixes #387.

### CODE REVIEW

**Comment by davmac314** [comment]
Thanks - I will try to review soon

File: src/dinit.cc [hunk 1]
```diff
@@ -718,7 +723,19 @@ int dinit_main(int argc, char **argv)
             // Services all stopped but there was no shutdown issued. Inform user, wait for ack, and
             // re-start boot sequence.
             sync(); // Sync to minimise data loss if user elects to power off / hard reset
-            confirm_restart_boot();
+            if (auto_recovery) {
+                try {
+                    log_msg_end(" Will start recovery.");
```
Comment (davmac314) [inline src/dinit.cc:728]: `log_msg_end` is supposed to be paired with `log_msg_begin` (and zero or more `log_msg_part`s). To issue a message that doesn't consist of multiple parts you can use a single call to `log` instead.

The message `" Will start recovery."` is not self-contained (and starts with a space). I think maybe you copied and modified from one of the lines above, which are all part of a two-part message. If you log a message, it should be self-contained and should not start with white space. I would suggest just `"Starting recovery."`.

File: src/dinit.cc [hunk 2]
```diff
@@ -718,7 +723,19 @@ int dinit_main(int argc, char **argv)
...
+                    services->start_service("recovery");
+                }
+                catch (...) {
+                    std::cout << "Unable to start recovery service.\n";
```
Comment (davmac314) [inline src/dinit.cc:732]: This should go to the log, I think.

File: src/dinit.cc [hunk 3]
```diff
@@ -718,7 +723,19 @@ int dinit_main(int argc, char **argv)
...
+                    confirm_restart_boot();
+                }
+            } else {
```
Comment (davmac314) [inline src/dinit.cc:735]: You don't really need the `else` clause, you can just have `confirm_restart_boot()` after the if.

I.e. rather than:

```
if (...) {
    try {
        ...
    }
    catch (...) {
        ...
        confirm_restart_boot();
    }
}
else {
    confirm_restart_boot();
}

```

You can have:

```
if (...) {
    try {
        ...
    }
    catch (...) {
        ...
    }
}
confirm_restart_boot();
```

The extra log message (" Will prompt user.") isn't really needed.

**Review by davmac314 — CHANGES_REQUESTED** [review]

File: src/dinit.cc (hunk 1 above)
Comment (JamiKettunen) [reply src/dinit.cc:728]: The space was to account for https://github.com/davmac314/dinit/blob/927bec0/src/dinit.cc#L686-L697 not handling `shutdown_type_t::NONE` as I saw the log contained the next thing on the same line after `No more active services.` without new line, should that be dealt with or left as-is?

File: src/dinit.cc (hunk 3 above)
Comment (JamiKettunen) [reply src/dinit.cc:735]: The extra log thing was again to end https://github.com/davmac314/dinit/blob/927bec0/src/dinit.cc#L684 with a newline, but perhaps `shutdown_type_t::NONE` could be handled right below or be left as-is anyway, it's just currently missing a linebreak for the next thing appended to log

Comment (davmac314) [reply src/dinit.cc:735]: I see, that actually makes sense (and I think in fact you're fixing an existing issue), though it's a bit hard to follow, so I might re-arrange things later. For now then this logic is good, however the `else` should be on the next line to match the code style (see CODE-STYLE).

File: src/dinit.cc (hunk 1 above)
Comment (davmac314) [reply src/dinit.cc:728]: (Ok, this is fine, see my other comment).

**Comment by davmac314** [comment]
One other thing - the man page (doc/manpages/dinit.8.m4) needs to be updated with the new option.

File: src/dinit.cc (hunk 3 above)
Comment (JamiKettunen) [reply src/dinit.cc]: I've made it more clear what the intention was by using a more generic `Will handle boot failure.` where the others already were.

**Review by davmac314 — APPROVED** [review]
Looks good. Thanks!

**Comment by davmac314** [comment]
I invite contributors to add themselves to the CONTRIBUTORS file if they wish. You are welcome to do this by opening another pull request with that change. Thanks again.
